oaipmh/__init__.py

```diff
@@ -2,5 +2,92 @@
+documentation.
+
+Our implementation relies on the OAI-PMH repository build into Opencast. Opencast publishes
+additional metadata about mediapackages using it's own XML schema. Conventially, these are
+published under the "matterhorn" metadata prefix.
+
+Repositories are configured via the :py:class:`oaipmh.models.Repository` model. A repository has
+at least a URL and may optionally have some authentication information.
+
+Once configured, new metadata records can be harvested using either the ``oaipmh_harvest``
+management command or the :py:class:`oaipmh.tasks.harvest_all_repositories` and
+:py:class:`oaipmh.tasks.harvest_repository` Celery tasks.
+
+When new metadata is harvested, a :py:class:`oaipmh.models.Record` object is created for each
+record in the repository. Additionally, a :py:class:`oaipmh.models.MatterhornRecord` object is
+created for each Opencast media package.
+
+When a new :py:class:`oaipmh.models.MatterhornRecord` object is created, any tracks which match the
+:py:class:`oaipmh.records.TRACK_TYPE` type will have :py:class:`oaipmh.models.Track` objects
+created. Similarly, :py:class:`oaipmh.models.Series` objects will be created for all Opencast
+series which have records.
+
+The :py:class:`oaipmh.models.Series` model has a ``playlist`` field which points to the
+:py:class:`mediapackages.models.Playlist` associated with the series. This field has to be set
+manually at the moment. When a new :py:class:`oaipmh.models.Track` object is added to the database
+and its associated series has a playlist, a new media item is created from the track's content and
+added to the playlist. If the series has no associated playlist or if the track is of the wrong
+type, no media item is created.
+
+In order to handle cases where, for example, a series gains a playlist after tracks have already
+been harvested, there is a "cleanup" task which can be run via either the ``cleanup`` management
+command or the :py:class:`oaipmh.tasks.cleanup` Celery task. This task will try to run the various
+object creation/media upload jobs which the database assert should be done but which have not. This
+is usually only required if the database is changed manually or if there is an error uploading a
+media item.
+
+Set up
+------
+
+To set up the OAI-PMH harvester, follow the following steps:
+
+1. Create a :py:class:`oaipmh.models.Repository` object in the database representing the OAI-PMH
+   repository which should be harvested from.
+
+2. Schedule harvesting either by having the ``oaipmh_harvest`` management command run via a cronjob
+   or schedule the :py:class:`oaipmh.tasks.harvest_all_repositories` Celery task. It is recommended
+   that this job runs regularly, perhaps every minute. If you want to make sure you never miss any
+   metadata updates, one can schedule a "fetch all records" harvest nightly.
+
+3. It is also worth scheduling the ``oaipmh_cleanup`` management command or
```

Comment:
If all the automation works (playlists created/associated automatically, etc) I can't see any reason why cleanup should need to be executed as a nightly cronjob (even less, every 5 minutes). Once the database is large enough, it will constantly hammer the db... For development is a different case...

Reply:
In principle the Matterhorn Record creation should never fail but the Track creation could fail f JWP has a hiccough. Neither track creation or record creation cleanup jobs touch records they are not supposed to unless a "full" cleanup is requested. The record creation cleanup task is mostly intended for situations where we change the Matterhorn record parsing in the future and we need a convenient way to re-parse the existing records.

I agree that in a perfect world we wouldn't need a cleanup task but there's no harm in running a non-full one nightly.